File: 백준/Silver/5014. 스타트링크/스타트링크.py
from collections import deque
F,S,G,U,D = map(int, input().split())

# F가 최대 높이
# S가 현재 위치
# G가 목표 위치
# U가 위로 U만큼 이동
# D가 아래로 D만큼 이동

visited = [-1] * (F + 1)
visited[S] = 0
queue = deque([S])


while queue :
    cur = queue.popleft()

    for nxt in [cur+U, cur - D] :
        if 0 < nxt <= F and visited[nxt] == - 1 :
            visited[nxt] = visited[cur] + 1
            queue.append((nxt))

if visited[G] == -1 :
    print('use the stairs')
else :
    print(visited[G])

File: 백준/Silver/5014. 스타트링크/스타트링크.java
import java.io.*;
import java.util.*;

public class Main {
    static int f;
    static int s;
    static int g;
    static int u;
    static int d;
    static int[] visited;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());
        f = Integer.parseInt(st.nextToken());
        s = Integer.parseInt(st.nextToken());
        g = Integer.parseInt(st.nextToken());
        u = Integer.parseInt(st.nextToken());
        d = Integer.parseInt(st.nextToken());

        visited = new int[f + 1];
        Arrays.fill(visited,-1);


        bfs();
    }
    static void bfs(){
        Queue<Integer> queue = new LinkedList<>();
        queue.offer(s);
        visited[s] = 0;

        while (!queue.isEmpty()){
            int cur = queue.poll();

            if (cur == g) {
                System.out.print(visited[g]);
                return;
            }
            if (cur + u <= f && visited[cur + u] == -1){
                visited[cur + u] = visited[cur] + 1;
                queue.offer(cur + u);
            }
            if (cur - d >= 1 && visited[cur - d] == -1){
                visited[cur - d] = visited[cur] + 1;
                queue.offer(cur - d);
            }
        }
        System.out.print("use the stairs");
    }
}
